### PR TITLE
migration: add deduplication

### PR DESCRIPTION
**Pull request by jma**

### CODE REVIEW

**Comment by PascalRepond** [comment]
## Tests

- REP/07.10: Status not color coded in detailed view for migration data: https://ilsdev.test.rero.ch/professional/migrations/records/format/migration_data/detail/-dUtZpIBx4caacpVioYZ?migration=cogne
- REP/07.10: Dedup interface, some scores are shown for automatic matches (by identifier)
image
- REP/07.10: Dedup interface, it would be nice to display the translation of the value for `document.type`

File: rero_ils/modules/deduplications/__init__.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2019-2023 RERO
```
Comment (Garfield-fr) [inline rero_ils/modules/deduplications/__init__.py:4]: 2024 ?

File: tests/deduplications/ui/conftest.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2019-2023 RERO
```
Comment (Garfield-fr) [inline tests/deduplications/ui/conftest.py:4]: 2024 ?

File: tests/deduplications/ui/test_deduplications_api.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2021-2023 RERO
```
Comment (Garfield-fr) [inline tests/deduplications/ui/test_deduplications_api.py:4]: 2024?

File: tests/migrations/api/test_migrations_rest.py [hunk 4]
```diff
@@ -37,7 +37,8 @@ def test_migrations_rest(
     assert res.status_code == 200
     assert res.json["hits"]["total"]["value"] == 1
     assert res.json["hits"]["hits"][0]["id"] == migration.meta.id
-    assert res.json["hits"]["hits"][0]["metadata"] == migration.to_dict()
+    # TODO: check dates
+    # assert res.json["hits"]["hits"][0]["metadata"] == jsonify(migration.to_dict()).json
```
Comment (Garfield-fr) [inline tests/migrations/api/test_migrations_rest.py:41]: ?

**Review by Garfield-fr — APPROVED** [review]

**Review by PascalRepond — APPROVED** [review]

File: pyproject.toml [hunk 5]
```diff
@@ -117,6 +117,8 @@ dnspython = ">2.4.2"
 cryptography = ">41.0.7"
 babel = ">2.14.0"
 rero-invenio-files = ">=1.0.0,<2.0.0"
+levenshtein = "^0.26.0"
```
Comment (rerowep) [inline pyproject.toml:120]: For futur updates it is better to not fix the version to one number.

File: pyproject.toml [hunk 6]
```diff
@@ -117,6 +117,8 @@ dnspython = ">2.4.2"
...
+unidecode = "^1.3.8"
```
Comment (rerowep) [inline pyproject.toml:121]: see above

**Comment by PascalRepond** [comment]
## Tests phase 2

- [ ] Add/modify fields to rescore and to dedup UI for differentiation:
    - [ ] use `title._text` instead of `sort_title` for rescore
    - [ ] `editionStatement`: simple edit distance, boost^3
    - [ ] `identifiedBy`: boost^5
    - [ ] `seriesStatement[0]._text[0].value`: boost^2
- [x] Arrows to browse candidates could be closer together
- [ ] Never use documents with `_masked:true` as candidates

File: rero_ils/modules/deduplications/api.py [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+
+        :param text1: str - first string value.
+        :param text2: str - second string value.
+        :returns: [0.0-1.0] - the score: 1 means a perfect match
+        """
+        score = ratio(
+            self.normalize(text1),
+            self.normalize(text2),
+        )
+        return score if score > 0.6 else 0.0
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:228]: use 'score_cutoff=0.6' as a parameter.

File: rero_ils/modules/deduplications/api.py [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+        """Compute a similarity score between two strings.
+
+        Strings are normalized and the score will be normalized between 0 an 1.
+        A score < 0.6 is replaced by 0.
+
+        :param text1: str - first string value.
+        :param text2: str - second string value.
+        :returns: [0.0-1.0] - the score: 1 means a perfect match
+        """
+        score = ratio(
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:224]: I would try using jaro_winkler() instead of ratio(), a few tests were convincing.

**Comment by PascalRepond** [comment]
## Tests phase 3

- [ ] REP: Dedup: modification time is displayed in the wrong timezone.
- [ ] REP: Dedup: identifiers, if both are empty, return None and don't score (otherwise documents that have no identifiers are too much favorited, example BR02206)

File: rero_ils/modules/deduplications/api.py [hunk 9]
```diff
@@ -0,0 +1,523 @@
...
+            if not prov:
+                return {}
+            to_return = set()
+            for d in prov[0].get("statement", []):
+                if d["type"] == "bf:Agent":
+                    for label in d["label"]:
+                        to_return.add(self.normalize(label["value"]))
+            return to_return
+
+        # build a dict with type: date
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:168]: Incorrect comment

File: rero_ils/modules/deduplications/api.py [hunk 10]
```diff
@@ -0,0 +1,523 @@
...
+        prov = get_editor(data)
+        new_candidates = []
+        for candidate in candidates:
+            # build a dict with type: date for each candidate
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:172]: Incorrect comment

File: rero_ils/modules/deduplications/api.py [hunk 11]
```diff
@@ -0,0 +1,523 @@
...
+        :param text1: str - first string value.
+        :param text2: str - second string value.
+        :returns: [0.0-1.0] - the score: 1 means a perfect match
+        """
+        score = jaro_winkler(text1, text2, score_cutoff=0.6, processor=cls.normalize)
+        return score
+
+    @classmethod
+    def get_title_score(cls, data, candidate):
+        """Get the title score.
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:258]: Compare and score the similarity of the first title's `_text` in both documents.

File: rero_ils/modules/deduplications/api.py [hunk 12]
```diff
@@ -0,0 +1,523 @@
...
+        :returns: the edit distance of the title value.
+        """
+        return cls.edit_distance(
+            main_title_text(data["title"])[0]["_text"],
+            main_title_text(candidate["title"])[0]["_text"],
+        )
+
+    @classmethod
+    def get_series_statement_score(cls, data, candidate):
+        """Get the provision activity score.
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:269]: Compare and score the similarity of the first seriesStatement's `_text` in both documents.

File: rero_ils/modules/deduplications/api.py [hunk 13]
```diff
@@ -0,0 +1,523 @@
...
+        return (
+            1
+            if cls.normalize(data_field[0]["_text"][0]["value"])
+            == cls.normalize(c_field[0]["_text"][0]["value"])
+            else 0
+        )
+
+    @classmethod
+    def get_edition_statement_score(cls, data, candidate):
+        """Get the edition statement score.
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:286]: Compare and score the similarity of the first editionStatement's `_text` in both documents.

File: rero_ils/modules/deduplications/api.py [hunk 14]
```diff
@@ -0,0 +1,523 @@
...
+        return (
+            1
+            if cls.normalize(data_field[0]["_text"][0]["value"])
+            == cls.normalize(c_field[0]["_text"][0]["value"])
+            else 0
+        )
+
+    @classmethod
+    def get_doc_type_score(cls, data, candidate):
+        """Get the document type score.
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:303]: Compare and score the document types in both documents.

File: rero_ils/modules/deduplications/api.py [hunk 15]
```diff
@@ -0,0 +1,523 @@
...
+            if (
+                {t["main_type"] for t in data["type"]}
+                & {t["main_type"] for t in candidate["type"]}
+            )
+            else 0
+        )
+
+    @classmethod
+    def get_publication_date_score(cls, data, candidate):
+        """Get the document type score.
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:318]: Compare and score the sort_date_old in both documents.

File: rero_ils/modules/deduplications/api.py [hunk 16]
```diff
@@ -0,0 +1,523 @@
...
+        publication_date = data.get("sort_date_old")
+        c_publication_date = candidate.get("sort_date_old")
+        if publication_date and c_publication_date:
+            return 1 if publication_date == c_publication_date else 0
+        else:
+            return 0 if bool(publication_date) != bool(c_publication_date) else 1
+
+    @classmethod
+    def get_provision_activity_score(cls, data, candidate):
+        """Get the provision activity score.
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:331]: Compare and score the similarity of the first provisionActivity's _text in both documents.

File: rero_ils/modules/deduplications/api.py [hunk 17]
```diff
@@ -0,0 +1,523 @@
...
+
+        :returns: 1 if both exist and are equal matched, 0 otherwise.
+        """
+        data_field = data["provisionActivity"][0]["_text"][0]["value"]
+        c_field = candidate["provisionActivity"][0]["_text"][0]["value"]
+        return cls.edit_distance(data_field, c_field)
+
+    @classmethod
+    def get_extent_score(cls, data, candidate):
+        """Get the provision activity score.
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:341]: Compare and score the similarity of the numerical values in both documents' `extent`.

File: rero_ils/modules/deduplications/api.py [hunk 18]
```diff
@@ -0,0 +1,523 @@
...
+        data_field = set(re.findall(r"\d+", data.get("extent", "")))
+        c_field = set(re.findall(r"\d+", candidate.get("extent", "")))
+        if data_field and c_field:
+            return len(data_field & c_field) / max(len(data_field), len(c_field))
+        else:
+            return 0 if bool(data_field) != bool(c_field) else 1
+
+    @classmethod
+    def get_responsibility_score(cls, data, candidate):
+        """Get the document type score.
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:354]: Compare and score the similarity of the first responsibilityStatement's `_text` in both documents.

File: rero_ils/modules/deduplications/api.py [hunk 19]
```diff
@@ -0,0 +1,523 @@
...
+        if resp and c_resp:
+            return cls.edit_distance(resp[0][0]["value"], c_resp[0][0]["value"])
+        else:
+            return 0 if bool(resp) != bool(c_resp) else 1
+
+    @classmethod
+    def get_identifier_score(cls, data, candidate):
+        """Get the document type score.
+
+        :returns: 1 if both exist and are equal matched.
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:369]: :returns: 1 if both documents have at least one identifier in common (excluding bf:Local).

File: rero_ils/modules/deduplications/api.py [hunk 20]
```diff
@@ -0,0 +1,523 @@
...
+        resp = data.get("responsibilityStatement")
+        c_resp = candidate.get("responsibilityStatement")
+        if resp and c_resp:
+            return cls.edit_distance(resp[0][0]["value"], c_resp[0][0]["value"])
+        else:
+            return 0 if bool(resp) != bool(c_resp) else 1
+
+    @classmethod
+    def get_identifier_score(cls, data, candidate):
+        """Get the document type score.
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:367]: Compare identifiers in both documents.

File: rero_ils/modules/deduplications/api.py [hunk 21]
```diff
@@ -0,0 +1,523 @@
...
+        )
+
+    def get_text_candidates(self, data):
+        """Get candidates using text queries (fuzzy search).
+
+        :param data: the data to import.
+        :returns: the list of the candidate.
+        """
+        # build the text query
+        # use the sort_title field as it contains the normalized title
```
Comment (PascalRepond) [inline rero_ils/modules/deduplications/api.py:450]: Delete this line

File: tests/deduplications/ui/test_deduplications_api.py [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+    assert score < 0.95
+
+    doc["responsibilityStatement"] = deepcopy(document_data["responsibilityStatement"])
+    doc["responsibilityStatement"][0].append({"value": "foo"})
+    doc.replace(doc, True, True, True)
+    [(pid, _, score, _)] = dedup.get_candidates(deepcopy(document_data))
+    assert (pid, score) == (doc.pid, 1.0)
+
+    # ============== text query ===============
+    # remove identifiers
```
Comment (PascalRepond) [inline tests/deduplications/ui/test_deduplications_api.py:99]: not needed ?

File: tests/deduplications/unit/test_deduplication_scores.py [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+    assert (
+        Deduplication.get_provision_activity_score(
+            document, create("Paris :  m. Levy, 1860")
+        )
+        == 1
+    )
+    assert (
+        Deduplication.get_provision_activity_score(document, create("Paris, 1860")) > 0
+    )
+
```
Comment (PascalRepond) [inline tests/deduplications/unit/test_deduplication_scores.py:167]: Maybe also test with a similar string "Pully : Marc levy, 1860" should be > 0

**Review by PascalRepond — CHANGES_REQUESTED** [review]

**Comment by PascalRepond** [comment]
Awesome job as a whole! 🥳

File: tests/deduplications/ui/test_deduplications_api.py (hunk 22 above)
Comment (jma) [reply tests/deduplications/ui/test_deduplications_api.py:99]: No as all the tests has been moved to the unit tests.

**Review by PascalRepond — APPROVED** [review]

**Pull request merged**